File: backend/apps/tickets/validators/basic_field_validator.py
"""
Chain of Responsibility node — validates basic text fields of a ticket.

Responsibility (SRP): enforce only character-count rules on asunto and descripcion.
Depends on: BaseValidator (core/base/base_validator.py).
Pattern: Chain of Responsibility node.
SOLID: SRP · OCP · LSP
"""

from __future__ import annotations

from core.base.base_validator import BaseValidator, ValidationResult


class BasicFieldValidator(BaseValidator):
    """Validates asunto (≤80 chars) and descripcion (≥10 chars)."""

    ASUNTO_MAX = 80
    DESCRIPCION_MIN = 10

    def validate(self, data: dict) -> ValidationResult:
        asunto = data.get("asunto", "")
        if not asunto or len(asunto) > self.ASUNTO_MAX:
            return ValidationResult(
                is_valid=False,
                errors=[f"El asunto debe tener entre 1 y {self.ASUNTO_MAX} caracteres."],
                field_name="asunto",
            )

        descripcion = data.get("descripcion", "")
        if not descripcion or len(descripcion) < self.DESCRIPCION_MIN:
            return ValidationResult(
                is_valid=False,
                errors=[
                    f"La descripción debe tener al menos {self.DESCRIPCION_MIN} caracteres."
                ],
                field_name="descripcion",
            )

        return ValidationResult(is_valid=True)

File: frontend/src/modules/tickets/validators/BasicFieldValidator.ts
/**
 * Chain of Responsibility node — validates text fields on the ticket form.
 *
 * Responsibility (SRP): enforce only character-count rules on asunto and descripcion.
 *     No file checks, no API calls — just field length.
 * Depends on: BaseValidator (src/core/base/BaseValidator.ts).
 * Pattern: Chain of Responsibility node.
 * SOLID: SRP · OCP · LSP
 *
 * Rules enforced:
 *   - asunto:      required, max 80 characters
 *   - descripcion: required, min 10 characters
 *
 * OCP: new field rule = new validator node; this class unchanged.
 *
 * Usage (assembled by ValidatorFactory):
 *   const chain = ValidatorFactory.buildTicketChain()
 *   const result = chain.run(formData)
 */

import { BaseValidator, type ValidationResult } from '../../../core/base/BaseValidator'

export class BasicFieldValidator extends BaseValidator {
  private static readonly ASUNTO_MAX = 80
  private static readonly DESCRIPCION_MIN = 10

  validate(data: unknown): ValidationResult {
    const { asunto = '', descripcion = '' } = data as { asunto?: string; descripcion?: string }
    if (!asunto.trim())
      return { isValid: false, field: 'asunto', errors: ['El asunto es requerido.'] }
    if (asunto.length > BasicFieldValidator.ASUNTO_MAX)
      return { isValid: false, field: 'asunto', errors: [`El asunto no puede superar ${BasicFieldValidator.ASUNTO_MAX} caracteres.`] }
    if (descripcion.trim().length < BasicFieldValidator.DESCRIPCION_MIN)
      return { isValid: false, field: 'descripcion', errors: [`La descripción debe tener al menos ${BasicFieldValidator.DESCRIPCION_MIN} caracteres.`] }
    return { isValid: true, field: '', errors: [] }
  }
}
